app/controllers/budgets/results_controller.rb

```diff
@@ -1,22 +1,31 @@
 module Budgets
   class ResultsController < ApplicationController
+    before_action :load_budget
+    before_action :load_heading
 
     load_and_authorize_resource :budget
 
     def show
       authorize! :read_results, @budget
       @investments = load_result.investments
-      @heading = heading
     end
 
     private
 
       def load_result
-        Budget::Result.new(@budget, heading)
+        Budget::Result.new(@budget, @heading)
       end
 
-      def heading
-        @budget.headings.find(params[:heading_id])
+      def load_budget
+        @budget = Budget.find_by(id: params[:budget_id])
+      end
+
+      def load_heading
+        if params[:heading_id].present?
```

Comment:
What do you think about replacing the whole function body with:
```
 @heading = @budget.headings.find_by(id: params[:heading_id]) || @budget.headings.first
```
When slugs come just replacing the `id:` for `slug:` will do